sources/healthdog/src/healthdog.rs

```diff
@@ -0,0 +1,161 @@
+/// to check the health of a list of services and send information about whether or not the services
+/// are running.
+pub(crate) struct Healthdog {
+    /// The `Healthdog` configuration, e.g. from `/etc/healthdog.toml`
+    config: Config,
+    /// Information about the Bottlerocket release, e.g. from `os-release`
+    os_release: BottlerocketRelease,
+    /// A trait object that checks if a service (listed in `config`) is healthy. This can be passed-
+    /// in, but defaults to an object that checks `systemd` services by name.
+    healthcheck: Box<dyn ServiceCheck>,
+}
+
+impl Healthdog {
+    /// Create a new instance by optionally passing in the `Config`, the `BottlerocketRelease`, and
+    /// `SystemCheck` objects. For each of these, if `None` is passed, then the default is used.
+    pub(crate) fn from_parts(
+        config: Option<Config>,
+        os_release: Option<BottlerocketRelease>,
+        healthcheck: Option<Box<dyn ServiceCheck>>,
+    ) -> Result<Self> {
+        Ok(Self {
+            config: match config {
+                None => Config::new()?,
+                Some(c) => c,
+            },
+            os_release: match os_release {
+                None => BottlerocketRelease::new().context(error::BottlerocketRelease)?,
+                Some(b) => b,
+            },
+            healthcheck: healthcheck.unwrap_or_else(|| Box::new(SystemdCheck {})),
+        })
+    }
+
+    /// # Description
+    ///
+    /// Sends key-value pairs as query parameters in a GET request to the URL in `config`. A
+    /// standard set of key-value pairs are sent first, and appended by any additional parameters
+    /// passed in to this function. The standard keys are: sender, event, version, variant, arch,
+    /// region, seed, version-lock, and ignore-waves.
+    ///
+    /// # Parameters
+    ///
+    /// * `sender`:          This is the name of the application sending the metrics e.g.
+    ///                      `healthdog` or `updog`.
+    /// * `event`:           The name of the type of metrics event that is being sent. For example
+    ///                      `boot-success` or `health-ping`.
+    /// * `values`:          The key-value pairs that you want to send. These will be sorted by key
+    ///                      before sending to ensure consistency or key-value ordering.
+    /// * `timeout_seconds`: The timeout setting for the HTTP client setting. Defaults to
+    ///                      `DEFAULT_TIMEOUT_SECONDS` when `None` is passed.
+    pub(crate) fn send<S1, S2>(
+        &self,
+        sender: S1,
+        event: S2,
+        values: Option<&HashMap<String, String>>,
+        timeout_seconds: Option<u64>,
+    ) -> Result<()>
+    where
+        S1: AsRef<str>,
+        S2: AsRef<str>,
+    {
+        let mut url = Url::from_str(&self.config.metrics_url).context(error::UrlParse {
+            url: self.config.metrics_url.clone(),
+        })?;
+        {
+            let mut q = url.query_pairs_mut();
```

Comment:
Might be a double mut

```suggestion
            let q = url.query_pairs_mut();
```